vendor/github.com/ethereum/go-ethereum/p2p/peer.go

```diff
@@ -248,7 +264,24 @@ func (p *Peer) pingLoop() {
 	}
 }
 
-func (p *Peer) readLoop(errc chan<- error) {
+func (p *Peer) heartbeatLoop(reads <-chan struct{}) {
+	defer p.wg.Done()
+	hb := time.NewTimer(2 * pingInterval)
+	defer hb.Stop()
+	for {
+		select {
+		case <-reads:
+			p.SetNonFlaky()
```

Comment:
don't we need to reset the watchdog there?

Reply:
yes, the idea is that if we were able to read from a socket then the connection is not flaky